### ISSUE TITLE
How to remove bundled resources per target?

### ISSUE THREAD

**Issue by thanpolas** [electron-userland/electron-builder#182]
In my project I am bundling binaries per target (darwin, win) so I want to delete the ones that aren't needed per target, when building osx delete the win binaries and vice versa...

Is there a way to intervene between the package stage and the dist stage so as to manipulate the package before it's bundled in an installable?

**Comment by develar** [comment]
* You bundle binaries with your app.
* Your app directory contains all binaries — for all OS.
* On package you want to include only binaries for target OS.

Right?

Our goal to provide you workflow, but not tons of options.
So, solution —

* Some directory will be considered as `resources`. Name it `resources`. Located in the root of project.
* This directory `resources` contain subdirectories per OS (osx, win, win64, linux, linux64).
* On package content of `resources/$platform-$arch` will be copied to `resources` in the packed app.

Is it a solution for you? If yes, could you please add your current naming and directories mappings? (where is your bundled binaries located and where do you copy it in the packaged app).

**Comment by thanpolas** [comment]
Yes, It's exactly like you say, except i didn't target arch, so my folder structure is just `/resources/darwin/` but I could use arch too...

As my understanding of the tools grows i've devised a way to accomplish that as well as avoid having the double package.json architecture...

I am copying the whole project onto a new folder and configure the packager/builder to use that folder, that way I can completely control what files / modules / folders get bundled, and with the help of the `prune` option get rid of dev deps, here's my `build` key of my package.json:

```json
{
  "build": {
    "arch": "x64",
    "dir": "out/app-copy/",
    "platform": "darwin",
    "app-bundle-id": "App",
    "app-category-type": "Sports",
    "app-copyright": "Acme, Inc. All Rights Reserved",
    "asar": false,
    "name": "App",
    "out": "dist/",
    "overwrite": true,
    "prune": true,
    "icon": "build/icon"
  }
}
```

It would be really cool if we could configure the `build/` folder files and not require to create a folder, or even worse move stuff around (as was in my case) because `build/` folder was already used

**Comment by develar** [comment]
> but I could use arch too...

Will be not required — if no `-64`, directory without arch prefix will be copied. 

> avoid having the double package.json architecture...

Hmm... You decided to solve inexistent issue. Ok, up to you.

> if we could configure the build/ folder files

You want ability to specify path to another folder?

**Comment by thanpolas** [comment]
> Hmm... You decided to solve inexistent issue. Ok, up to you.

It's not non-existent, bundling dev dependencies is an issue. Two package.json is one solution, using `prune` is another... unless i'm not understanding the problem correctly...

> You want ability to specify path to another folder?

Yes, be able to configure the `build/` folder alltogether, or even better specific locations per file (background.png, and icon.***, they're just two files)

**Comment by develar** [comment]
> prune is another.

No. It is not a solution, but awkward workaround :) In case of Two package.json you don't need to prune on each build (and then reinstall dev deps), you don't need to rebuild native deps on each build (and then rebuild dev deps).

**Comment by thanpolas** [comment]
The flow for prune is:

1. Copy the whole app directory into a target folder
1. Perform any granular controls on what goes into the bundle by deleting test folder etc (how is this managed today?)
1. Use that new folder for packager and builder and apply the `prune` option on...

If native apps have already been built on your original app, then there's no need to rebuild native deps...

The outcome is exactly the same (correct me if i'm wrong please)... The difference is that the `prune` way might take a bit longer... However I feel more in control since I have the chance to granularly shape what files will be bundled per build target... If you show me how I can have the same power without the "copy app to new folder" solution i'd really help me understand how to use these tools

**Comment by develar** [comment]
> Perform any granular controls on what goes into the bundle by deleting test folder etc (how is this managed today?)

`app` folder must contains only and only app data. Nothing more. `test` is not part of the app, so, must be located in the root of project (alongside development package.json).
You don't need do granular controls — just put in the `app` exactly what you want to be part of the app and nothing more.

`electron-builder` can and should do smart filtering — bundled binaries for target platform only, for example.

**Comment by thanpolas** [comment]
How is the `app/` folder activated? As long as I have an `app/` folder the builder will use that instead of the whole app? I don't really get that part... how does it work?

For me, were I am porting an existing application into electron, it'll be a big challenge to move everything under `app/`...

**Comment by develar** [comment]
>  I don't really get that part... how does it work?

If `app` folder exists in the project, builder will automatically switch to two-package.json layout. And use `app/package.json` as an app package.json and `./package.json` as a dev package.json.

See https://github.com/develar/onshape-desktop-shell
In the `app` you see only `package.json`. Because typescript compiler will compile TS files to JS to `app/out` — https://github.com/develar/onshape-desktop-shell/blob/master/tsconfig.json#L7

**Comment by develar** [comment]
> I am porting an existing application into electron

Do you mean that existing application will continue to be non-electron only, and, so, your team doesn't approve moving files?

**Comment by thanpolas** [comment]
> Do you mean that existing application will continue to be non-electron only, and, so, your team doesn't approve moving files?

No, there will be no backwards support, we're 100% electron from here on... but moving hundreds of files and updating paths will be quite some work...

In that scenario where we have all files under `app/` where would the `resouces/` folder go which has the bundled binaries per platform? I understand it should be within `app/` so the application can access it after the build operation

**Comment by develar** [comment]
> In that scenario where we have all files under app/ where would the resouces/ folder go which has the bundled binaries per platform?

Yes, you understand correctly — since it is a part of app, " it should be within app/ so the application can access it after the build operation". `electron-builder` can (this issue about it :)) and will soon filter out according to my [proposal](https://github.com/loopline-systems/electron-builder/issues/182#issuecomment-189148442).

> there will be no backwards support, we're 100% electron from here on... but moving hundreds of files and updating paths will be quite some work

I strongly advice do it. Or.... Well, to be honest, if you problem is only and only "get rid of dev dependencies", we can implement it very easy — it is not a rocket science to read your `package.json` and filter out all non-production dependencies. Without prune. Just copy only production deps to app dist. 

But! In the reality you have to get rid of `test` and other non-app files. Because your root project directory contains a lot of unrelated files.

Right? So, question is — are you ready to begin constant battle? Do you want to maintain exclude rules? And what will you do, when your app will required use of native dependencies (e.g. keytar, nslog)?

It is the reason why we are not going to implement simple feature "do not copy dev deps". Because a lot of issues must be solved before we can say — you can use one-package layout and your app will not contain unrelated junk.

**Comment by thanpolas** [comment]
> Right? So, question is — are you ready to begin constant battle? Do you want to maintain exclude rules?

I'd like that decision to be made by the app author and not a library... Moving everything into `app/` is a very radical move that will require a lot of assessment before it's decided, i'm not against it, i'm just point out that this is not trivial.

There's still the issue of production npm dependencies that need pruning as well. Our app is in a mature state with over 80 prod deps... Those 80 prod deps also need pruning as they contain their test folders, sample folders and docs, they amount to possibly hundreds of megabytes. So I'm afraid maintaining exclude rules and deletion scripts is a necessity.

I believe that having the ability to process the Packaged app `bundle.app` before starting the build operation to create the .dmg file is essential in the build flow. This could be accomplished if there was a distinct building command that accepted a `bundle.app` as input, vs coupling the two operations together (package & build).

**Comment by thanpolas** [comment]
...and then, there are the cases of the built files and their sources, I am talking about sass/less, js and their templates sources that get bundled into single files and minimized.

These sources need to be deleted.

**Comment by develar** [comment]
>  I am talking about sass/less, js and their templates sources that get bundled into single files and minimised.

Sources is not part of app. Example: in the https://github.com/develar/onshape-desktop-shell `src` located outside of `app` folder.  Compiler then will compile it to `app/out`.

> I'd like that decision to be made by the app author and not a library.

We don't dictate. `builder` supports one-package.json layout. I just say, that we are not going to solve obvious issues with this layout.

**Comment by thanpolas** [comment]
I see...

So is there anyway to accomplish the operation decoupling so as to process the packaged app `bundle.app` before creating the `bundle.dmg` distributable?

**Comment by develar** [comment]
> So is there anyway to accomplish the operation decoupling so as to process the packaged app

Will be implemented (ability to provide own js implementation to process out dir). But I guess for your `ignore` electron packager option is a working solution (I cannot give a direct link, find it on the https://github.com/maxogden/electron-packager page). (except filtering out binaries).

**Comment by develar** [comment]
@thanpolas Have you found a solution? Does `ignore` electron packager option works for you? Also, please see https://github.com/electron-userland/electron-builder/issues/230#issuecomment-196007808.

**Comment by thanpolas** [comment]
I've opted to use directly the appdmg package to create my bundle for so i could overcome the blocks i had, so far things look good, thank you for your help!

**Comment by develar** [comment]
@thanpolas If you still want to use electron-builder and it is a show-stopper for you —feel free to reopen issue.

Closed for now as won't fix — PR is welcome or I will try to implement solution if it will be required by you/another user.

**Comment by develar** [comment]
You can use https://github.com/electron-userland/electron-builder/wiki/Options#BuildMetadata-afterPack
